c10/core/TensorImpl.h

```diff
@@ -129,6 +129,20 @@ struct C10_API PlacementDeleteContext {
   }
 };
 
+// One TensorImpl can be copied to another TensorImpl if they have the same
+// type_id. The only two special cases (for legacy reason) are:
+// CPUTensorId is compatible with CUDATensorId and SparseCPUTensorId is
+// compatible with SparseCUDATensorId.
+inline bool has_compatible_type_for_copy(
+    const TensorTypeId& self,
+    const TensorTypeId& other) {
+  return (self == other) ||
+      ((self == CPUTensorId() || self == CUDATensorId()) &&
+      (other == CPUTensorId() || other == CUDATensorId())) ||
+      ((self == SparseCPUTensorId() || self == SparseCUDATensorId()) &&
+      (other == SparseCPUTensorId() || other == SparseCUDATensorId()));
```

Comment:
I guess this would technically break complex cpu/cuda tensors, but that's probably okay.

Reply:
@dzhulgakov @yf225 can comment? I don't have enough context to determine whether we should add complex cpu/cuda pair here... (according to dzhulgakov the current two pairs are for legacy reason so adding them here as special rule instead of some more generic condition based on is_sparse/is_cpu/is_cuda to cover more similar cases).

Reply:
I believe the full story is that in prior releases there was an extra layer of indirection so doing:
`tensor.data = other` was legal for any tensor types for both `tensor` and `other`.

There's no longer that extra level of indirection, so we changed the support to only allow you to do this if it works to do a shallow copy (i.e. same TensorImpl type).  We haven't done a release with this behavior.

Furthermore, complex types is an extension (i.e. you have to install it separately), so in theory someone could be doing the above with `tensor` as CPU complex and `other` as CUDA complex, but that seems pretty unlikely given all the moving pieces here.

Reply:
@gchanan - thanks for the context. yeah - I wasn't sure how far we'd like to go to limit cross-type shallow-copy. I'm using strict type comparison + 2 white-list cases here which is more strict than already-tightened-same-TensorImpl-type constraint. It makes more sense if the 2-white-list cases are relatively 'fixed' (they are allowed only for some legacy reason). If we need keep extending it as we introduce new type (as @bddppq commented below for HIPTensorId case) I feel it's a little bit suboptimal.
@dzhulgakov your thoughts? do you feel HIPTensorId is the only case we should add or there will be more legit cases we need add here?